OVERVIEW.md

```diff
@@ -0,0 +1,483 @@
+# Pseudocode
+
+The following attempts to capture the general logic of the Automation OCR2 
+implementation using pseudocode. It is divided first by the OCR2 interface 
+followed by descriptions of secondary services.
+
+## How an Observation is Constructed
+
+An automation OCR2 observation is constructed of a block number and a shuffled
+list of upkeep IDs (big.Int), json encoded to a limit of max bytes. The list
+should only include eligible upkeeps (covered in a later section). Shuffling
+is done using a randomness source derived from the config digest, epoch, and
+block number. This ensure all nodes have the same source, but the results are
+random across different OCR rounds if an observation is repeated.
+
+> the shuffling should not be necessary in the observation because we 
+> cannot trust that shuffling was done once the observation is consumed in Report
+
+```
+ocr -> PluginThatImplementsLibOCR2
+poller -> ParallelServicePollingLogs
+filter -> ParallelServiceWatchingUpkeepLogs
+
+call ocr.Observation with [ReportTimestamp]
+    call poller.SampleUpkeeps with filter.Filter(): return [blockNumber] and [upkeepStateResults]
+
+    define [upkeepIDList]
+    for each (upkeepStateResults)
+        if (upkeepState) is Eligible
+            add (upkeepID) to (upkeepIDList)
+        
+    call shuffle with (upkeepIDList) and [randomSourceFromReportTimestamp]
+
+    define [observation]
+    add (blockNumber) to (observation)
+    
+    define [observationBytes]
+    for each (upkeepIDList)
+        add (upkeepID) to (observation)
+        call encode with (observation): return [bytes]
+            if length of (bytes) < [maxObservationLength]
+                set (observationBytes) to (bytes)
+                continue
+            else
+                remove (upkeepID) from (observation)
+                call encode with (observation): return [bytes]
+                set (observationBytes) to (bytes)
+                break
+
+    return (observationBytes)
+```
+
+## How a Report is Constructed
+
+Observations from all other nodes (including the one processing a report) are
+processed into a final encoded report ready to be transmitted on chain. Key
+checks include:
+
+- all observations must pass decoding
+- a report must be built off of at least 1 observation
+- all observations must have a block number to calculate median from
+- observation id length should be less than or equal to max
+- check keys should be less than max keys for a report
```

Comment:
what are the check keys here? Upkeep IDs?

Reply:
updated with more explanation